Check if file exists before downloading - GH #125

cdkini opened the pull request:
I've added the following changes per our discussion in #125:
- Added force_redownload flag as optional arg
- Added check for existing path to prevent duplicate downloads of corporum (unless flag is turned on)
- Updated docstring to reflect new arg and exception

pytest on "audiomate/test/corpus/io/" yields 8 failures both before and after changes. Please let me know what you think whenever you have a chance!

ynop commented:
Thanks looks good.

* What do you mean with failures?
* Could you add some tests?
    * Files are deleted if flag is set
    * Exception is thrown before downloading data if flag is not set

cdkini commented:
The failures were all FileNotFoundErrors but I think the fault lies with me. I didn't set up the project using a venv and I think my dependencies were off. There are no errors present in "audiomate/test/corpus/io/" per my most recent run of pytest.
I've added tests to the following files: 

- test_aed.py 
- test_downloader.py
- test_esc.py
- test_free_spoken_digits.py
- test_gtzan.py
- test_librispeech.py
- test_mailabs.py
- test_musan.py
- test_rouen.py
- test_tatoeba.py
- test_voxforge.py

To test if files were deleted when the flag was set to true, I added a temp file and checked for its existence after rerunning the download method. To test if an exception was properly thrown when the flag was not set, I simply reran the method on an already populated target directory. I heavily leveraged the test_download methods in each file when writing these. 

Let me know if this fits what you were looking for or if I can make any other changes.

ynop commented:
I would go another way:
* You now test the same thing multiple times. I would do it only once for the base class. Create `tests/corpus/io/test_base.py` and create a mock downloader (extending `CorpusDownloader`), that just creates the directory and a file.
* At the moment you are testing two scenarios in one test. I would split them up. (e.g. one test `force_redownload_overwrites_existing_directory` and one test `existing_directory_forces_io_error`). If something fails you directly know which functionality does not work and it is easier to find errrors.

cdkini commented:
Okay thanks for the feedback that definitely makes more sense than my prior approach. 

I've created `test_base.py` and include the discussed tests therein. I wrapped `MockCorpusDownloader` in a helper function because it was being erroneously picked up by `CorpusDownloader.__subclasses__()` and causing issues with `test_helpers.py`. All checks pass with this approach but let me know if it can be improved at all.

ynop commented:
Looks good, could you merge your commits into one? I will merge it afterwards.

cdkini commented:
It should now all be squashed to 324df37c33477eedb445f81d53a6b6cd1ab1c957
